examples/flower_classifier/keras_image_classifier.py

```diff
@@ -0,0 +1,378 @@
+from keras.applications import *
+from keras.applications import vgg16
+from keras.layers import Input, Dense, Flatten, Lambda
+
+import mlflow
+import mlflow.pyfunc
+import mlflow.keras
+from mlflow.utils.file_utils import TempDir
+from mlflow.utils.environment import _mlflow_conda_env
+
+
+def read_image(filename):
+    with open(filename, "rb") as f:
+        return f.read()
+
+
+def decode_and_resize_image(raw_bytes, size):
+    return np.asarray(Image.open(BytesIO(raw_bytes)).resize(size), dtype=np.float32)
+
+
+class MLflowLogger(Callback):
+    """
+    Logs training metrics and final model with MLflow.
+
+    We log metrics provided by Keras during training and keep track of the best model (best loss
+    on validation dataset). At the end of the training, we log the model with weights that produced
+    the best result on the validation dataset during training.
+
+    The output model accepts raw images on the output and produces class label on the output when
+    scored. It is a composition of the actual DL model with following pre and post processing:
+       - preprocessing: decode and resize the image.
+       - post-processing: select output class id
+    """
+
+    def __init__(self, model, x_train, y_train, x_valid, y_valid,
+                 **kwargs):
+        self._model = model
+        self._best_val_loss = math.inf
+        self._train = (x_train, y_train)
+        self._valid = (x_valid, y_valid)
+        self._pyfunc_params = kwargs
+        self._best_weights = None
+
+    def on_epoch_end(self, epoch, logs=None):
+        """
+        Log Keras metrics with MLflow. If model improved on the validation data, evaluate it on
+        a test set and store it as the best model.
+        """
+        if not logs:
+            return
+        for name, value in logs.items():
+            if name.startswith("val_"):
+                name = "valid_" + name[4:]
+            else:
+                name = "train_" + name
+            mlflow.log_metric(name, value)
+        val_loss = logs["val_loss"]
+        if val_loss < self._best_val_loss:
+            # Save the "best" weights
+            self._best_val_loss = val_loss
+            self._best_weights = [x.copy() for x in self._model.get_weights()]
+
+    def on_train_end(self, logs=None):
+        self._model.set_weights(self._best_weights)
+        x, y = self._train
+        train_res = self._model.evaluate(x=x, y=y)
+        for name, value in zip(self._model.metrics_names, train_res):
+            mlflow.log_metric("train_{}".format(name), value)
+        x, y = self._valid
+        valid_res = self._model.evaluate(x=x, y=y)
+        for name, value in zip(self._model.metrics_names, valid_res):
+            mlflow.log_metric("valid_{}".format(name), value)
+        KerasImageClassifierPyfunc.log_model(keras_model=self._model, **self._pyfunc_params)
+
+
+class KerasImageClassifier(object):
+    """
+        Base class for image classification with MLflow and Keras.
```

Comment:
Use a consistent style for these doc comments; I think the comment text is supposed to be aligned with the `"""`, not indented underneath it.